### ISSUE TITLE
reactor semantics of `stop-when` not right?

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#808]
Here's a program:
```
  r = reactor:
    init: pts,
    stop-when: {(ps): is-empty(ps)},
    on-tick: {(ps): ps.rest},
    to-draw: {(ps): place-image(BALL,
            (ps.first + 1) * SCALE,
            50,
          BG)}
  end
```
The reactor state is a list of values. Given that the `stop-when` has a guard for emptiness, I'd expect `on-tick` and `to-draw` to not have to deal with empty lists. Yet I'm getting an error at `ps.first` saying that it's being applied to the empty list.

(On a practical note, it's difficult to write this program for the empty list, because it's just not clear what to do other than signal an error.)

The problem seems to only be with to-draw. The following program runs properly to completion:
```
r2 = reactor:
  init: [list: 1, 2, 3],
  stop-when: {(ps): is-empty(ps)},
  on-tick: {(ps): ps.rest}
end

r2.interact()
```
So it looks like `on-tick` is called after the `stop-when` check, but `to-draw` is called before it?

**Comment by sorawee** [comment]
FWIW:

```
(big-bang (list 1 2 3)
          (stop-when empty?)
          (to-draw (lambda (x) (circle (first x) "solid" "red")))
          (on-tick rest))
```

fails with `first: expects a non-empty list; given: '()`

**Comment by jpolitz** [comment]
I think you're pointing out the motivation for `last-picture`, @shriram.  At least up until now, I've always understood it that reactors (worlds) try to draw the first frame for which stop-when returns true.

**Comment by sorawee** [comment]
`last-picture` wouldn't help here though if we want to show the picture before `stop-when` returns `true`. Take my code above as an example, `last-picture` would consume an empty list, because that's the state where `stop-when` returns `true`. However, I think what Shriram would expect would be a circle of size 3. This is already gone and not recoverable.

**Comment by shriram** [comment]
Oak, no. This is the other benefit to stop-when, as Joe says.  If that
function wanted to do something based on the history of the computation, it
would have to enhance the state. For instance, if the final scene wanted to
show your game score, it would have to put the score in the state. If the
game ended even before it began, then your score world be whatever the init
value of it.

**Comment by shriram** [comment]
Sorry, last-picture, not stop-when.

--

**Comment by shriram** [comment]
This also provides a nice partitioning of states and preconditions. During
computation ones assume the stop-when condition is false. Last-picture
assumes it's true.

--
Sent from phone. Please pardon terseness and mistakes.

**Comment by jpolitz** [comment]
Is the conclusion that we should add `last-picture`, and that will handle this case?

Or should we add `last-picture`, _and also_ make `to-draw` not be called on stopped states, instead staying with the rendering of the previous state?

**Comment by shriram** [comment]
It's not enough to add `last-picture`. I view `last-picture` as a nice feature but not essential. It feels essential for the reactor to not call handlers when the stopping invariant has been met (equivalently, when the liveness invariant has been violated). I feel the program I wrote is a very natural one — stop when it's empty, react this way when it's not — and failure to honor this results in having to handle states that are unnatural. Worse, right now it appears `to-draw` and `on-tick` diverge in their handling!

In conclusion: check first for `stop-when`. If it is false, call a `last-picture` handler if provided with that state. If it is true, then call the other handlers.

**Comment by schanzer** [comment]
Tracking - @ds26gte is looking into adding this feature upstream from Patch.

**Comment by blerner** [comment]
@shriram I think your conclusion line above is backwards, right?  If there's a `stop-when` and it returns _true_, then call `last-picture` if there is one.  If there's not `stop-when` handler, or if it returns _false_, then continue with the `on-tick` and `to-draw` handlers.

**Comment by shriram** [comment]
Yes, I have it backwards. What you wrote above is correct.
